compiler/backend/src/org/jetbrains/kotlin/codegen/state/KotlinTypeMapper.kt

```diff
@@ -383,7 +383,7 @@ class KotlinTypeMapper @JvmOverloads constructor(
         resolvedCall: ResolvedCall<*>? = null
     ): CallableMethod {
         // we generate constructors of inline classes as usual functions
-        if (descriptor is ConstructorDescriptor && kind !== OwnerKind.ERASED_INLINE_CLASS) {
+        if (descriptor is ConstructorDescriptor && (kind !== OwnerKind.ERASED_INLINE_CLASS || isIrBackend)) {
```

Comment:
I don't think it makes sense to completely separate the two here, since there is only this single special case in which they differ. On the other hand, it definitely makes sense to separate the name mangling rules, since this is much simpler for the IrBackend.